File: qiskit/circuit/library/evolution/suzuki_trotter.py
# This code is part of Qiskit.
#
# (C) Copyright IBM 2021.
#
# This code is licensed under the Apache License, Version 2.0. You may
# obtain a copy of this license in the LICENSE.txt file in the root directory
# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
#
# Any modifications or derivative works of this code must retain this
# copyright notice, and modified files need to carry a notice indicating
# that they have been altered from the originals.

"""The Lie-Trotter product formula."""

from typing import List, Callable, Optional, Union
import numpy as np
from qiskit.circuit.quantumcircuit import QuantumCircuit
from qiskit.quantum_info.operators import SparsePauliOp, Pauli

from .product_formula import ProductFormula


class SuzukiTrotter(ProductFormula):
    """The (higher order) Suzuki-Trotter product formula."""

    def __init__(
        self,
        reps: int = 1,
        order: int = 2,
        atomic_evolution: Optional[Callable[[SparsePauliOp, float], QuantumCircuit]] = None,
        insert_barriers: bool = False,
        cx_structure: str = "chain",
    ) -> None:
        """
        Args:
            order: The order of the product formula.
            reps: The number of time steps.
            atomic_evolution: A function to construct the circuit for the evolution of single operators.
                Per default, `PauliEvolutionGate` will be used.
        """
        super().__init__(order, reps, atomic_evolution, insert_barriers, cx_structure)

    def synthesize(
        self, operators: Union[SparsePauliOp, List[SparsePauliOp]], time: float
    ) -> QuantumCircuit:
        if not isinstance(operators, list):
            pauli_list = [(Pauli(op), np.real(coeff)) for op, coeff in operators.to_list()]
        else:
            pauli_list = [(op, 1) for op in operators]

        ops_to_evolve = self._recurse(self.order, time / self.reps, pauli_list)

        single_rep = QuantumCircuit(operators[0].num_qubits)
        first_barrier = False

        for op, coeff in ops_to_evolve:
            # add barriers
            if first_barrier:
                if self.insert_barriers:
                    single_rep.barrier()
            else:
                first_barrier = True

            single_rep.compose(self.atomic_evolution(op, coeff), inplace=True)

        evo = QuantumCircuit(operators[0].num_qubits)
        first_barrier = False

        for _ in range(self.reps):
            # add barriers
            if first_barrier:
                if self.insert_barriers:
                    single_rep.barrier()
            else:
                first_barrier = True

            evo.compose(single_rep, inplace=True)

        return evo

    @staticmethod
    def _recurse(order, time, pauli_list):
        if order < 1:
            raise ValueError("This bitch empty -- yeet!")

        if order == 1:
            return pauli_list

        elif order == 2:
            halves = [(op, coeff * time / 2) for op, coeff in pauli_list[:-1]]
            full = [(pauli_list[-1][0], time * pauli_list[-1][1])]
            return halves + full + list(reversed(halves))

        else:
            reduction = 1 / (4 - 4 ** (1 / (order - 1)))
            outer = 2 * SuzukiTrotter._recurse(
                order - 2, time=reduction * time, pauli_list=pauli_list
            )
            inner = SuzukiTrotter._recurse(
                order - 2, time=(1 - 4 * reduction) * time, pauli_list=pauli_list
            )
            return outer + inner + outer

File: crates/circuit_library/src/suzuki_trotter.rs
// This code is part of Qiskit.
//
// (C) Copyright IBM 2025
//
// This code is licensed under the Apache License, Version 2.0. You may
// obtain a copy of this license in the LICENSE.txt file in the root directory
// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
//
// Any modifications or derivative works of this code must retain this
// copyright notice, and modified files need to carry a notice indicating
// that they have been altered from the originals.

use crate::pauli_evolution::{Instruction, sparse_term_evolution};
use qiskit_circuit::Qubit;
use qiskit_circuit::circuit_data::CircuitData;
use qiskit_circuit::operations::{Param, StandardInstruction, multiply_param, radd_param};
use qiskit_circuit::packed_instruction::PackedOperation;
use qiskit_quantum_info::sparse_observable::SparseObservable;
use qiskit_synthesis::evolution::suzuki_trotter::{evolution, reorder_terms};
use smallvec::smallvec;
use thiserror::Error;

pub fn suzuki_trotter_evolution(
    observable: &SparseObservable,
    order: u32,
    reps: u32,
    time: f64,
    preserve_order: bool,
    insert_barriers: bool,
) -> Result<CircuitData, EvolutionError> {
    if order > 1 && order % 2 != 0 || order == 0 {
        return Err(EvolutionError::OrderSymmetry(order));
    }

    let terms_iter = observable.iter().map(|mut view| {
        view.coeff.re *= time * 2.0 / (reps as f64);
        view
    });
    let terms = if preserve_order || !preserve_order && observable.bit_terms().len() <= 1 {
        terms_iter.collect()
    } else {
        match reorder_terms(terms_iter) {
            Ok(terms) => terms,
            Err(msg) => return Err(EvolutionError::TermsReorder(msg.to_string())),
        }
    };

    // execute evolution
    let evo: Vec<(usize, f64)> = evolution(order, terms.len());

    // convert terms into instructions for circuit creation
    let repetitions = evo.len() as u32 * reps;
    let iter_repeated = evo.iter().cycle().take(repetitions as usize);

    let mut global_phase = Param::Float(0.0);
    let mut modified_phase = false;
    for (index, coeff) in iter_repeated.clone() {
        let view = &terms[*index];
        if view.bit_terms.len() == 0 {
            global_phase = radd_param(global_phase, (view.coeff.re * coeff).into());
            modified_phase = true;
        }
    }

    let repeated_evo = iter_repeated
        .enumerate()
        .map(|(i, (index, coeff))| {
            let view = &terms[*index];
            let instructions = sparse_term_evolution(
                view.bit_terms
                    .iter()
                    .map(|bit| bit.py_label())
                    .collect::<String>()
                    .leak(),
                view.indices.into(),
                (view.coeff.re * coeff).into(),
                false,
                false,
            )
            .map(Ok::<Instruction, _>);

            let maybe_barrier = (insert_barriers && i as u32 != repetitions - 1)
                .then_some(Ok(create_barrier(observable.num_qubits())))
                .into_iter();
            instructions.chain(maybe_barrier)
        })
        .flatten();

    if modified_phase {
        global_phase = multiply_param(&global_phase, -0.5);
    }

    match CircuitData::from_packed_operations(
        observable.num_qubits() as u32,
        0,
        repeated_evo,
        global_phase,
    ) {
        Ok(circuit) => Ok(circuit),
        Err(_) => Err(EvolutionError::CircuitBuild),
    }
}

fn create_barrier(num_qubits: u32) -> Instruction {
    (
        PackedOperation::from_standard_instruction(StandardInstruction::Barrier(num_qubits)),
        smallvec![],
        (0..num_qubits as u32).map(Qubit).collect(),
        vec![],
    )
}

#[derive(Debug, Error)]
pub enum EvolutionError {
    /// An invalid order value passed to the main method
    #[error["Suzuki product formulae are symmetric and therefore only defined \
            for when the order is 1 or even, not {0}."]]
    OrderSymmetry(u32),

    /// A general error when trying to build circuit from generated instructions
    #[error["Failed building circuit"]]
    CircuitBuild,

    /// A general error for terms reordering error
    #[error["Error ocurred when trying to reorder terms: {0}"]]
    TermsReorder(String),
}
